### DISCUSSION TITLE
Should we release 0.68.1?

### DISCUSSION THREAD

**Discussion by kelset** [reactwg/react-native-releases#19] [Patches]
RELEASED - please refer to https://github.com/reactwg/react-native-releases/discussions/20 for 0.68.2

---

Current Release: [0.68.0](https://github.com/facebook/react-native/releases/tag/v0.68.0)
    
Conversations on this thread are limited: 
* [major release issues](https://github.com/facebook/react-native/wiki/Release-FAQ#what-is-release-blocking)
* [qualified cherry-pick requests](https://github.com/facebook/react-native/wiki/Release-FAQ#what-is-a-qualified-pick-request) of commits on main that [did not make the previous patch version](https://github.com/facebook/react-native/wiki/Release-FAQ#how-do-i-know-if-my-fixfeature-is-in-a-certain-release).
    
Please include a link to the specific commit on main to be cherry-picked, for example: https://github.com/facebook/react-native/commit/20b0eba581a00e5e7e300f6377379b836617c147
    
In other words, if you cannot point to a particular commit on main, then your request likely belongs as a new issue. 
If the issue is a [major release issue](https://github.com/facebook/react-native/wiki/Release-FAQ#what-is-release-blocking), please reference the issue here.
    
---
#### List of qualified picks
 
1. https://github.com/facebook/react-native/commit/58a2eb7f37c2dc27ad3575618778ad5b23599b27
1. https://github.com/facebook/react-native/commit/5dff920177220ae5f4e37c662c63c27ebf696c83
1. https://github.com/facebook/react-native/commit/5534634892f47a3890e58b661faa2260373acb25
1. https://github.com/facebook/react-native/commit/44ee801c2980998b7be2543e3e40e281519329a6
1. https://github.com/facebook/react-native/commit/4befd2a29cb94b026d9c048a041aa9f1817295b5
1. https://github.com/facebook/react-native/commit/e48a580080bdae58b375f30fbcf8a83cc1915b2f 
1. https://github.com/facebook/react-native/pull/33581 Which includes the following fixes for for react-native-gradle-plugin https://github.com/facebook/react-native/commit/9573d7b84d35233fbb39a4067cfef65490aa34a7 (version bump of `react-native-gradle-plugin`) https://github.com/facebook/react-native/commit/48113b5635270cfbcef804f78ba0fb0b1aa4ca14 (update `yarn.lock` with the version)
    1. https://github.com/facebook/react-native/commit/d5da70e17e8c8210cd79a4d7b09c6a5ded4b5607
    1. https://github.com/facebook/react-native/commit/34628ac12ae3609153d0da22f676f25b8f9158a4

#### Local commits to backport to main
1.

**Comment by kelset** [comment]
as per conversation with @douglowder, we should cherry pick this: https://github.com/facebook/react-native/commit/58a2eb7f37c2dc27ad3575618778ad5b23599b27 for reanimated to work correctly with the new arch

**Comment by kelset** [comment]
We want to address this: https://github.com/facebook/react-native/issues/33525

**Reply by cortinico** [reply]
Kudos to @danilobuerger for sending this super quick fix:
- https://github.com/facebook/react-native/pull/33530

I'm working on making this land.
The trick would be that we'll have to cherry-pick a bump of the `react-native-gradle-plugin` package (plus we need a point release for that NPM package, which is something that we'll have to do manually).

**Reply by tido64** [reply]
Since you're already cherry-picking a bump, can you also remove `react-native-codegen` from dependencies?

**Reply by cortinico** [reply]
Yup that already landed: https://github.com/facebook/react-native/commit/34628ac12ae3609153d0da22f676f25b8f9158a4

**Reply by danilobuerger** [reply]
Commit is at https://github.com/facebook/react-native/commit/d5da70e17e8c8210cd79a4d7b09c6a5ded4b5607

**Reply by cortinico** [reply]
This has been included inside https://github.com/facebook/react-native/pull/33581
RNGP 0.0.6 is now on NPM and contains the two fixes. I've verified that it works correctly with a RN 0.68 project so can safely be backported.

**Comment by matinzd** [comment]


**Reply by danilobuerger** [reply]
was already mentioned above

**Reply by danilobuerger** [reply]
That probably doesn't qualify, see https://github.com/facebook/react-native/wiki/Release-FAQ#what-is-a-qualified-pick-request

**Reply by matinzd** [reply]
Makes sense. I will delete the conversation.

**Comment by myckhel** [comment]
**Dropping incase this would be useful:**

After upgrading to rn v0.68.0 and you came across this error `Build input file cannot be found: '**/ios/**/AppDelegate.m'` and you are the type of person that ignores collapsed `project.pbxproj` from [upgrade-helper](https://react-native-community.github.io/upgrade-helper/), just know that you might have omitted a very important change from that file which changed `AppDelegate.m` to `AppDelegate.mm`.

### In order to fix this
> Open the project in XCode, then navigate to Build Phases, in Compile Sources remove the AppDelegate.m file and add AppDelegate.mm instead.

**Reply by kelset** [reply]
I checked and in the upgrade-helper it shows up like so:
<img width="558" alt="Screenshot 2022-04-01 at 10 20 13" src="https://user-images.githubusercontent.com/16104054/161280408-cc452d60-5040-43c5-b212-870e071f99ad.png">

I think that's a good enough indication of the change, but thanks for adding the comment here! Let's see over time if more folks hit your same issue

**Reply by leotm** [reply]
It's indicated at the top too:
![Capture](https://user-images.githubusercontent.com/1881059/161721712-47b1ffbd-97c7-42a1-a71f-f79a26768198.PNG)

Just in case still missed I've [suggested ](https://github.com/react-native-community/upgrade-helper/pull/306) a comment pointing to [iOS: Use Objective-C++ (.mm extension)](https://reactnative.dev/docs/next/new-architecture-app-intro#ios-use-objective-c-mm-extension)

**Reply by Adnan-Bacic** [reply]
im on xcode 13.3 for reference

when i add the AppDelegate.mm file through the steps you mentioned, it gives me a dialog.

<img width="727" alt="Screenshot 2022-04-05 at 13 37 59" src="https://user-images.githubusercontent.com/32088941/161745708-6cbeba67-382d-4aef-8658-316e2fe9b5cc.png">

do we just keep the default boxes checked? or do we change anything

**Reply by myckhel** [reply]
Well for me i kept them checked but i didn't understand why i should or not.

**Comment by danilobuerger** [comment]
We should look into https://github.com/facebook/react-native/issues/33555 which is a regression caused by the introduction of bundler / ruby version. This should be reworked as both methods of installing cocoapods (gem / homebrew) are incorrect, and it should use bundler instead. The linked issue is caused by the fact that gem install of cocoapods is not invoked in the rn project dir but one up and so it uses a different ruby version.

**Reply by mikehardy** [reply]
I would love to see that stuff fixed. I just commented there with my experience. It's not fair to hate on code, it's not the code's fault, and it was put in for a great reason (I collaborated a bit on that) but I *hate* the DX now that it's in.

**Reply by danilobuerger** [reply]
@mikehardy maybe we can also check if there is actually any reason to use ruby 2.7 instead of the current 3.1?

**Reply by mikehardy** [reply]
I'm not aware of any reason personally - I've been using ruby-current-stable consistently as it comes out for ages now with only one hiccup (a serious one though) so in general it should work to just use current stable. But I am sure there are some things that for some reasons work on 2.x not 3.x so it's probably more of a compatibility "band" we should verify (e.g., >= 2.7.x actually is needed). My issue is that *enforcing* a specific ruby version, especially an old one, for no reason is to me not correct. The goal with the original work was to rely on known versions so compatibility would be assured, a worthy goal, I did not realize at the time it wasn't an optional thing if implemented the way it was

**Reply by kelset** [reply]
It sounds to me that there are two branching actionables from here:
* a quick fix for the 0.68 could be to remove the `.ruby-version` file from the template (or change it so that >=2.7 everything is valid), and tell people to do the same if they are hitting issues
* a "chonkier" fix for main branch which both addresses the current DX and updates Ruby to current stable release (3.1.x)

Am I missing anything?

**Reply by danilobuerger** [reply]
I personally would like to keep ruby-version and bump ruby to 3.1.

I have an open PR ( https://github.com/facebook/react-native/pull/33485 ) bumping ruby to 2.7.5, we could use that and bump to 3.1 instead.

**Reply by kelset** [reply]
the thing is that would be a fairly impactful change (most likely requiring folks to bump their ruby dep from 2.x to 3.x) so I'm not really into the idea of having that happen on a patch release of 0.68. I'm all for it for the next version, but I'd like to have something less impactful for 0.68.1

**Reply by danilobuerger** [reply]
How so? If you use a stock ruby without a ruby version manager, the `.ruby-version` doesn't affect you afaik. If you are using a ruby version manager, it will just install that version for that specific directory and everywhere else it will be what was already set globally. What will affect stock ruby is the `Gemfile`, but there we could (and should!) use a range specifier instead (>= 2.7 as that is what cocoapods needs currently).

**Reply by kelset** [reply]
> What will affect stock ruby is the Gemfile, but there we could (and should!) use a range specifier instead (>= 2.7 as that is what cocoapods needs currently).

ok and this can be done within your PR too? Sorry I'm not a master of Ruby, I'm just trying to wrap my head around what changes are "breaking" vs what is just "relaxing" existing boundaries

**Reply by danilobuerger** [reply]
Yeah so if we set `.ruby-version` to `3.1.1` the ruby version managers should just pick that up and it will only be valid for that folder and subfolder. Depending on the version manager it requires a command to install that version if it was not previously installed. For example rbenv will error out with:

> rbenv: version `3.1.1' is not installed

which is then fixed by 

```bash
❯ rbenv install
```

Normal ruby without a version manager ignores `.ruby-version` afaik.

For the `Gemfile`, `bundler` will complain if the version is a mismatch:

```bash
❯ bundle install
Your Ruby version is 2.7.4, but your `Gemfile` specified 3.1.1
```

Thats why I would set the ruby version requirement in the `Gemfile` to `>= 2.6` which is the min requirement for `cocoapods`. This would then even work with the mac stock ruby version without having to install a different one.

**Reply by kelset** [reply]
thanks for the great explanation! That sounds like a perfect approach 👍

**Reply by mikehardy** [reply]
I think cocoapods is >= 2.7, otherwise, if this works, fantastic until 3.1.2 comes out, then we'll be back to the start.
What really would happen if we just deleted .ruby-version ? (and why doesn't ruby version allow range specification, I tried and it did not work for me - would solve the whole thing if we could just provide unbounded upper end)

**Reply by danilobuerger** [reply]
> I think cocoapods is >= 2.7,

From https://blog.cocoapods.org/CocoaPods-1.11.0-beta/ :

> CocoaPods 1.11 raises the minimum Ruby version to 2.6

I don't think having the .ruby-version around is such a big problem. When you don't have a ruby version manager installed, this file is simply ignored. If you have one, it will use that version only for that folder and subfolders. If x.y.z comes out and we still use 3.1.1 it doesn't affect any other project on your machine. And once we validate that x.y.z works we can bump it. (I believe that was the intention of the initial inclusion of this.)

But yes, not much will change if we simply remove that file. I think the inclusion of the Gemfile (and therefor the version range of cocoapods) is far more important than the ruby version which seems to have always worked fine (except that one time that might have triggered the inclusion of those files with ffi and m1).

Maybe it was just a bit premature without having the docs and `react-native init` updated?

**Comment by douglowder** [comment]
For the `AppDelegate.mm` change I just edited the `project.pbxproj` file and did a global replace of `AppDelegate.m` -> `AppDelegate.mm`... worked great for me

**Comment by danilobuerger** [comment]
We might need https://github.com/facebook/react-native/commit/5534634892f47a3890e58b661faa2260373acb25
This keeps the `use-subscription` dependency under v1.6.0 which in turn started to depend on react 18.

**Reply by kelset** [reply]
yeah good shout, I've seen a couple folks getting hit by this 👍

**Reply by fortmarek** [reply]
Thanks @danilobuerger - I've added it to the list of commits to cherry-pick 👌

**Comment by kelset** [comment]
The fix for the Windows + Android + New Arch is being worked on here: https://github.com/facebook/react-native/pull/33582 I think it'd be good to see it merged for the 0.68 branch and then do 0.68.1.
Then a separate conversation will have to be had on how to backport this to main and/or if it's just a matter of "reverse cherry pick" it - I'll let @cortinico and @mganandraj deliberate on that ;)

**Reply by cortinico** [reply]
I'd say for this we should be fine merging this on `main` + cherry-pick on 0.68

**Comment by cortinico** [comment]
I've worked on a fix for M1 users on Android with the New Architecture:
- https://github.com/facebook/react-native/pull/33588

This will make sure users can build on M1s without having to patch their environments (i.e. edit the `ndk-build` file) in any form. We might want to include this into a patch for 0.68.x

EDIT: Clarify that this is for new arch

**Reply by yotamishak** [reply]
Thank you for this 🙏

**Comment by jspizziri** [comment]
Would it be possible to get https://github.com/facebook/react-native/commit/dac56ce0776e0e4d23ed4f8b324f2e2432aefa6a cherry-picked and into this patch?

**Reply by fortmarek** [reply]
Hey @jspizziri 👋 

This does not seem as a regression in 0.68. Is there a specific reason we'd want this included in 0.68.1 already or can it wait for 0.69?

**Reply by jspizziri** [reply]
Hey @fortmarek , thanks for the response. First time looking into the RN release process, but I did read the literature provided via links in the comment descriptions and I think I just misunderstood: [`"Is a fix for an issue introduced in current release."`](https://reactnative.dev/contributing/release-faq#what-is-a-qualified-pick-request). This particular issue doesn't seem to have been a regression between 0.67 & 0.68, so it sounds like it's not a candidate for a patch under that criteria.

Sorry for the confusion!

**Reply by fortmarek** [reply]
Hey @jspizziri! Yeah, the phrasing is not entirely clear, sorry for that 🙇  I would then wait with this for 0.69. You can always patch this and we'll also soon start iterating on the 0.69 release.

**Reply by jspizziri** [reply]
@fortmarek thanks for the help! FYI, [I created an MR to clear up the language](https://github.com/facebook/react-native-website/pull/3071/files) in the event that it's helpful.
